src/benchmarks/src/end_to_end_bench.cpp

```diff
@@ -1,227 +1,261 @@
-        query_vec = planes_vecs[plane_index];
-        binary_labels.emplace_back(SKLEARN_PLANE_LABEL);
-      } else {
-        // Send a bird
-        size_t bird_index = std::rand() % birds_vecs.size();
-        query_vec = birds_vecs[bird_index];
-        binary_labels.emplace_back(SKLEARN_BIRD_LABEL);
+    for (int i = 0; i < request_batch_size; i++) {
+      query_num = j * request_batch_size + i;
+      query_vec = data[query_num % num_datapoints];
+
+      if (prevent_cache_hits) {
+        // Modify it to be epoch and thread-specific
+        query_vec[0] = query_num / num_datapoints;
+        query_vec[1] = thread_id;
       }
 
-      std::shared_ptr<Input> cifar_input =
-          std::make_shared<DoubleVector>(query_vec);
-      boost::future<Response> future =
-          qp.predict({TEST_APPLICATION_LABEL,
-                      0,
-                      cifar_input,
-                      100000,
-                      clipper::DefaultOutputSelectionPolicy::get_name(),
-                      {VersionedModelId(SKLEARN_MODEL_NAME, "1")}});
-      futures.push_back(std::move(future));
-    }
+      std::shared_ptr<Input> input = std::make_shared<DoubleVector>(query_vec);
+      Query q = {TEST_APPLICATION_LABEL,
+                 UID,
+                 input,
+                 latency_objective,
+                 clipper::DefaultOutputSelectionPolicy::get_name(),
+                 {VersionedModelId(model_name, model_version)}};
 
-    std::shared_ptr<std::atomic_int> completed =
-        std::make_shared<std::atomic_int>(0);
-    std::pair<boost::future<void>, std::vector<boost::future<Response>>>
-        results = future::when_all(std::move(futures), completed);
-    results.first.get();
-    for (int i = 0; i < static_cast<int>(results.second.size()); i++) {
-      boost::future<Response> &future = results.second[i];
-      double label = static_cast<double>(binary_labels[i]);
-      double pred = std::stod(future.get().output_.y_hat_);
-      if (pred == label) {
-        accuracy_ratio->increment(1, 1);
-      } else {
-        accuracy_ratio->increment(0, 1);
-      }
+      boost::future<Response> prediction = qp.predict(q);
+      bench_metrics.request_throughput_->mark(1);
+
+      prediction.then([bench_metrics](boost::future<Response> f) {
+        Response r = f.get();
+
+        // Update metrics
+        if (r.output_is_default_) {
+          bench_metrics.default_pred_ratio_->increment(1, 1);
+        } else {
+          bench_metrics.default_pred_ratio_->increment(0, 1);
+        }
+        bench_metrics.latency_->insert(r.duration_micros_);
+        bench_metrics.num_predictions_->increment(1);
+        bench_metrics.throughput_->mark(1);
+      });
     }
-    std::this_thread::sleep_for(std::chrono::milliseconds(batch_delay_millis));
+
+    delay_micros =
+        draw_from_poisson ? distribution(generator) : batch_delay_micros;
+    std::this_thread::sleep_for(std::chrono::microseconds(delay_micros));
```

Comment:
That is good to know -- we can just skip the sleep code if the `delay_micros` is <= 0